### PR TITLE
Add new `Naming/PrivateMethodName` cop

### PR DESCRIPTION
**Pull request by kyoshidajp**
This PR adds new `Naming/PrivateMethodName` cop. It checks that private methods names start with a underscore prefix.

```ruby
# bad
class MyClass
  private
  def foo
    # ...
  end
end

# bad
class MyClass
  def self.foo
    # ...
  end

  private_class_method :foo
end

# good
class MyClass
  private
  def _foo
    # ...
  end
end

# good
class MyClass
  def self._foo
    # ...
  end

  private_class_method :_foo
end
```

Sometimes there are cases where private methods are prefixed with an underscore (e.g. [Active Model](https://github.com/rails/rails/blob/v7.0.7.2/activemodel/lib/active_model/callbacks.rb#L128-L153)). This is a good idea because it makes it easier to determine if a method is private or not.

The following topics may need to be considered.

1. `protected` methods are should be covered?
   - Actually, I am not sure which one it should be since I rarely use the protected method. It might be better without it.
1. `attr` accessors are should be covered?
   - If covered, it also affects instance variables naming. I implemented it anyway.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/naming/private_method_name.rb [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Naming
+      # Checks that private methods names
+      # do not start without a underscore prefix.
```
Comment (bbatsov) [inline lib/rubocop/cop/naming/private_method_name.rb:7]: Might be good to mention here that the convention is used to some extent (e.g. in Rails), but it's not widely adopted in the community and that's the reason why this cop is disabled by default.

File: lib/rubocop/cop/naming/private_method_name.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      #   class MyClass
+      #     def self._foo
+      #       # ...
+      #     end
+      #
+      #     private_class_method :_foo
+      #   end
+      #
+      class PrivateMethodName < Base
+        extend AutoCorrector
```
Comment (bbatsov) [inline lib/rubocop/cop/naming/private_method_name.rb:45]: I guess you should remove this, as the cop doesn't implement auto-correct.

**Comment by bbatsov** [comment]
Overall I'm fine with the proposed cop, given that it will be disabled. I've left a bit of feedback. @koic Feel free to take a look as well.

**Comment by koic** [comment]
Starting private method names with an underscore sometimes indicates a struggle to find a good name for internal methods in contrast to public methods. In these cases, a more descriptive naming would be the essential solution.

It's questionable whether this rule should be a general guideline supported by RuboCop. For instance, there are numerous private methods without an underscore in the Rails repository.
https://github.com/rails/rails/blob/v7.0.7.2/activemodel/lib/active_model/dirty.rb#L250-L292

**Comment by kyoshidajp** [comment]
@koic Thank you for your comment.

I believe my suggestion is only optional and should be disabled by default. Certainly there are many private methods without underscores in several projects. I believe that this cop is effective in determining if a method is private or not, but I am not trying to enforce it. I hope that can be determined by each project.

Would you still be negative about adopting it?

**Comment by kyoshidajp** [comment]
OK. I do not wish to force the introduction of subtle rules that you do not agree with. Again, thank you for your comments and review.

**Pull request closed without merging**